File: src/ipc/movies.ts
import { invoke } from "@tauri-apps/api/core";
import { Movie, MovieFilter, MovieSort, MovieWithMedia } from "../types";

export async function listMovies(
  filter?: MovieFilter,
  sort?: MovieSort
): Promise<Movie[]> {
  return await invoke<Movie[]>("list_movies", { filter, sort });
}

export async function getMovie(movieId: string): Promise<MovieWithMedia | null> {
  return await invoke<MovieWithMedia | null>("get_movie", { movieId });
}

export async function searchMovies(query: string): Promise<Movie[]> {
  return await invoke<Movie[]>("search_movies", { query });
}

export async function setMetadataMatch(
  movieId: string,
  providerId: string
): Promise<Movie> {
  return await invoke<Movie>("set_metadata_match", { movieId, providerId });
}

File: src-tauri/src/commands/movies.rs
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tauri::State;
use uuid::Uuid;

use crate::domain::{Media, Movie, MovieFilter, MovieSort};
use crate::error::AppError;
use crate::services::{LibraryService, SearchService};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MovieWithMedia {
    pub movie: Movie,
    pub media: Vec<Media>,
}

#[tauri::command]
pub async fn list_movies(
    filter: Option<MovieFilter>,
    sort: Option<MovieSort>,
    library_service: State<'_, Arc<LibraryService>>,
) -> Result<Vec<Movie>, AppError> {
    let f = filter.unwrap_or_default();
    let s = sort.unwrap_or_default();
    library_service.list_movies(&f, s)
}

#[tauri::command]
pub async fn get_movie(
    movie_id: String,
    library_service: State<'_, Arc<LibraryService>>,
) -> Result<Option<MovieWithMedia>, AppError> {
    let id = Uuid::parse_str(&movie_id).map_err(|e| AppError::Validation(e.to_string()))?;
    match library_service.get_movie_with_media(&id)? {
        Some((movie, media)) => Ok(Some(MovieWithMedia { movie, media })),
        None => Ok(None),
    }
}

#[tauri::command]
pub async fn search_movies(
    query: String,
    search_service: State<'_, Arc<SearchService>>,
) -> Result<Vec<Movie>, AppError> {
    search_service.search(&query)
}

#[tauri::command]
pub async fn set_metadata_match(
    movie_id: String,
    provider_id: String,
    library_service: State<'_, Arc<LibraryService>>,
) -> Result<Movie, AppError> {
    let id = Uuid::parse_str(&movie_id).map_err(|e| AppError::Validation(e.to_string()))?;
    library_service.set_manual_metadata(&id, &provider_id)
}
